### This script contains:

#### 1. Import libraries & dataset
#### 2. Flags for Station uses
#### 3. Querying low Station uses
#### 4. Removing all Station ID's with less than 100 hires
#### 5. Splitting out Unique Stations

## 1. Import libraries & dataset

In [22]:
import quandl
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
import os


In [23]:
#Ensuring that the graphs you create are displayed within the notebook without the need to "call" them specifically.
%matplotlib inline

In [24]:
#Folder path into usable string
path = r'C:\Users\willm\Dropbox\1 Data Analytics Course\1 New York Citibike Hire'

In [25]:
#Import Hire Dataset
NYB2020_Stations = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'NYB2020merged.pkl'))

In [26]:
#Set Pandas to show all columns
pd.set_option("display.max_columns", None)

## 2. Flags for Station uses

Still using the NYB2020merged.pkl from Script 1, as I would like ALL uses - including the long ones - to be used.

In [27]:
NYB2020_Stations.head()

tripduration                 starttime                  stoptime  \
0           789  2020-01-01 00:00:55.3900  2020-01-01 00:14:05.1470   
1          1541  2020-01-01 00:01:08.1020  2020-01-01 00:26:49.1780   
2          1464  2020-01-01 00:01:42.1400  2020-01-01 00:26:07.0110   
3           592  2020-01-01 00:01:45.5610  2020-01-01 00:11:38.1550   
4           702  2020-01-01 00:01:45.7880  2020-01-01 00:13:28.2400   

   start_station_id               start_station_name  start_station_latitude  \
0               504                  1 Ave & E 16 St               40.732219   
1              3423  West Drive & Prospect Park West               40.661063   
2              3687                  E 33 St & 1 Ave               40.743227   
3               346              Bank St & Hudson St               40.736529   
4               372        Franklin Ave & Myrtle Ave               40.694546   

   start_station_longitude  end_station_id           end_station_name  \
0               -73.981656             307      Canal St & Rutgers St   
1               -73.979453            3300  Prospect Park West & 8 St   
2               -73.974498             259    South St & Whitehall St   
3               -74.006180             490            8 Ave & W 33 St   
4               -73.958014            3637    Fulton St & Waverly Ave   

   end_station_latitude  end_station_longitude  bikeid    usertype  \
0             40.714275             -73.989900   30326  Subscriber   
1             40.665147             -73.976376   17105    Customer   
2             40.701221             -74.012342   40177  Subscriber   
3             40.751551             -73.993934   27690  Subscriber   
4             40.683239             -73.965996   32583  Subscriber   

   birth_year  gender  start_hour start_date2  temp  rain_snow day_of_week  \
0        1992       1           0  2020-01-01     3        0.0   Wednesday   
1        1969       1           0  2020-01-01     3        0.0   Wednesday   
2        1963       1           0  2020-01-01     3        0.0   Wednesday   
3        1980       1           0  2020-01-01     3        0.0   Wednesday   
4        1982       1           0  2020-01-01     3        0.0   Wednesday   

   day_of_week_number    month  month_number  TripMins  age  
0                   3  January             1      13.0   28  
1                   3  January             1      26.0   51  
2                   3  January             1      24.0   57  
3                   3  January             1      10.0   40  
4                   3  January             1      12.0   38

In [28]:
#Creating a new column to allow me to GroupBy the Station IDs columns
NYB2020_Stations['1col'] = "1"

In [29]:
NYB2020_Stations['1col'] = NYB2020_Stations['1col'].astype('int8')

In [30]:
#Creating a GroupBy to give total uses by each 'age' in a new column.
NYB2020_Stations["uses_StartStation"] = NYB2020_Stations.groupby("start_station_id")['1col'].transform("sum")

In [31]:
NYB2020_Stations["uses_EndStation"] = NYB2020_Stations.groupby("end_station_id")['1col'].transform("sum")

In [32]:
NYB2020_Stations[['start_station_id', 'uses_StartStation']].head(20)

start_station_id  uses_StartStation
0                504            59142.0
1               3423            48873.0
2               3687            73560.0
3                346            38629.0
4                372             8071.0
5                394            28740.0
6               3093            33556.0
7                514            97415.0
8                274            21312.0
9                467            18195.0
10               504            59142.0
11              3374            52390.0
12               514            97415.0
13               536            40448.0
14               502            39244.0
15              3046             6928.0
16               514            97415.0
17              3709            45925.0
18              3349            40286.0
19               516            19817.0

Stations 514 shows up 3 times, with the same uses.

In [33]:
NYB2020_Stations[['end_station_id', 'uses_EndStation']].head(20)

end_station_id  uses_EndStation
0              307          52211.0
1             3300          16094.0
2              259          31685.0
3              490          62776.0
4             3637          16257.0
5              478          62586.0
6             3668          15573.0
7              426         101767.0
8              416          13779.0
9             3368          17005.0
10             393          26097.0
11            3379          14350.0
12             405          45905.0
13             174          35672.0
14             296            666.0
15            3068           6580.0
16             405          45905.0
17             440          26037.0
18            3411          10950.0
19             367          30044.0

Station 405 shows up twice, with the same uses.

## 3. Querying low Station uses

In [34]:
#Checking to see how many times Stations are used as a starting point - max and min
NYB2020_Stations['start_station_id'].value_counts(dropna = False, ascending=True)

4332         1
4310         2
4328         3
3912         3
4306         6
         ...  
499      91856
514      97415
435      99191
426      99364
3141    100753
Name: start_station_id, Length: 1213, dtype: int64

In [36]:
NYB2020_Stations['start_station_idSTR'] = NYB2020_Stations['start_station_id'].astype('str')

In [38]:
NYB2020_StartStation4332 = NYB2020_Stations[NYB2020_Stations['start_station_idSTR']=='4332']

In [39]:
NYB2020_StartStation4332.head()

tripduration                 starttime                  stoptime  \
19504238           318  2020-12-31 20:17:25.1670  2020-12-31 20:22:43.9950   

          start_station_id     start_station_name  start_station_latitude  \
19504238              4332  Morris Ave & E 163 St                40.82723   

          start_station_longitude  end_station_id       end_station_name  \
19504238               -73.917656            4169  Morris Ave & E 171 St   

          end_station_latitude  end_station_longitude  bikeid    usertype  \
19504238              40.83899               -73.9109   48231  Subscriber   

          birth_year  gender  start_hour start_date2  temp  rain_snow  \
19504238        1984       2          20  2020-12-31     7        1.0   

         day_of_week  day_of_week_number     month  month_number  TripMins  \
19504238    Thursday                   4  December            12       5.0   

          age  1col  uses_StartStation  uses_EndStation start_station_idSTR  
19504238   36     1                1.0             75.0                4332

<font color='blue'>First used on the last day of our data!</font>

In [40]:
NYB2020_StartStation4310 = NYB2020_Stations[NYB2020_Stations['start_station_idSTR']=='4310']

In [41]:
NYB2020_StartStation4310.head()

tripduration                 starttime                  stoptime  \
19484524           342  2020-12-31 12:01:46.9430  2020-12-31 12:07:29.3590   
19488691           264  2020-12-31 13:47:57.8940  2020-12-31 13:52:22.6310   

          start_station_id start_station_name  start_station_latitude  \
19484524              4310     24 St & 38 Ave               40.756351   
19488691              4310     24 St & 38 Ave               40.756351   

          start_station_longitude  end_station_id end_station_name  \
19484524               -73.936515            3586   34 Ave & 13 St   
19488691               -73.936515            3745   36 Ave & 31 St   

          end_station_latitude  end_station_longitude  bikeid    usertype  \
19484524             40.763875             -73.936485   46577  Subscriber   
19488691             40.756513             -73.929575   48559  Subscriber   

          birth_year  gender  start_hour start_date2  temp  rain_snow  \
19484524        1986       1          12  2020-12-31     7        1.0   
19488691        1985       1          13  2020-12-31     7        1.0   

         day_of_week  day_of_week_number     month  month_number  TripMins  \
19484524    Thursday                   4  December            12       6.0   
19488691    Thursday                   4  December            12       4.0   

          age  1col  uses_StartStation  uses_EndStation start_station_idSTR  
19484524   34     1                2.0           3354.0                4310  
19488691   35     1                2.0          10599.0                4310

<font color='blue'>First used on the last day of our data!</font>

In [42]:
NYB2020_StartStation3912 = NYB2020_Stations[NYB2020_Stations['start_station_idSTR']=='3912']

In [43]:
NYB2020_StartStation3912.head()

tripduration                 starttime                  stoptime  \
161165            278  2020-01-06 10:15:25.1070  2020-01-06 10:20:03.1980   
214920            486  2020-01-07 14:31:17.8780  2020-01-07 14:39:24.1440   
2551333          2063  2020-03-04 13:24:20.8520  2020-03-04 13:58:43.8620   

         start_station_id start_station_name  start_station_latitude  \
161165               3912        MTL-AOS-5.1                 40.7903   
214920               3912        MTL-AOS-5.1                 40.7903   
2551333              3912        MTL-AOS-5.1                 40.7903   

         start_station_longitude  end_station_id end_station_name  \
161165                  -73.9597            3912      MTL-AOS-5.1   
214920                  -73.9597            3912      MTL-AOS-5.1   
2551333                 -73.9597            3912      MTL-AOS-5.1   

         end_station_latitude  end_station_longitude  bikeid    usertype  \
161165                40.7903               -73.9597   37965  Subscriber   
214920                40.7903               -73.9597   37965  Subscriber   
2551333               40.7903               -73.9597   37965    Customer   

         birth_year  gender  start_hour start_date2  temp  rain_snow  \
161165         1996       1          10  2020-01-06     5        2.0   
214920         1996       1          14  2020-01-07     5        0.0   
2551333        1969       0          13  2020-03-04    11        0.0   

        day_of_week  day_of_week_number    month  month_number  TripMins  age  \
161165       Monday                   1  January             1       5.0   24   
214920      Tuesday                   2  January             1       8.0   24   
2551333   Wednesday                   3    March             3      34.0   51   

         1col  uses_StartStation  uses_EndStation start_station_idSTR  
161165      1                3.0              3.0                3912  
214920      1                3.0              3.0                3912  
2551333     1                3.0              3.0                3912

<font color='blue'>MTL-AOS-5.1's Long and Lat put it in the middle of Central Park, and all 3 trips started and ended there.</font>

## 4. Removing all Station ID's with less than 100 hires
#### Most will be Bike Stations that were launched later in the year, so have not had the time to accrue hires like other stations.

In [ ]:
NYB2020_Stations.info()

In [49]:
NYB2020_Stations['uses_StartStation'] = NYB2020_Stations['uses_StartStation'].astype('int64')

In [50]:
#Creating a new column to display one of two flags, depending on the age of the person hiring the bike
NYB2020_Stations.loc[NYB2020_Stations['uses_StartStation'] > 100, 'Over_100'] = 'Y'
NYB2020_Stations.loc[NYB2020_Stations['uses_StartStation'] <= 100, 'Over_100'] = 'N'

In [53]:
NYB2020_Stations[['start_station_id', 'uses_StartStation', 'Over_100']].head(20)

start_station_id  uses_StartStation Over_100
0                504              59142        Y
1               3423              48873        Y
2               3687              73560        Y
3                346              38629        Y
4                372               8071        Y
5                394              28740        Y
6               3093              33556        Y
7                514              97415        Y
8                274              21312        Y
9                467              18195        Y
10               504              59142        Y
11              3374              52390        Y
12               514              97415        Y
13               536              40448        Y
14               502              39244        Y
15              3046               6928        Y
16               514              97415        Y
17              3709              45925        Y
18              3349              40286        Y
19               516              19817        Y

In [54]:
NYB2020_Stations['Over_100'].value_counts(dropna = False, ascending=True)

N        1000
Y    19505857
Name: Over_100, dtype: int64

In [55]:
NYB2020_Over100Uses = NYB2020_Stations[NYB2020_Stations['Over_100']=='Y']

## 5. Splitting out Unique Stations

In [58]:
#Creating a subset of the unique Start Stations
NYB2020_StationsUniqueStarts = NYB2020_Over100Uses.drop_duplicates(subset = ["start_station_id"])


In [59]:
#Showing that each station only shows up once
NYB2020_StationsUniqueStarts['start_station_id'].value_counts(dropna = False, ascending=True)

504     1
3404    1
3407    1
3105    1
278     1
       ..
3359    1
3100    1
3780    1
3782    1
4280    1
Name: start_station_id, Length: 1189, dtype: int64

In [20]:
NYB2020_StationsUniqueStarts.to_csv(os.path.join(path, '02 Data','Prepared Data', 'NYB2020_UsesByStation.csv'), index = False)

tripduration                 starttime                  stoptime  \
19506837           148  2020-12-31 23:55:42.2030  2020-12-31 23:58:10.4400   
19506838            76  2020-12-31 23:56:20.4840  2020-12-31 23:57:37.2340   
19506839          1504  2020-12-31 23:56:28.5940  2021-01-01 00:21:33.2290   
19506840           268  2020-12-31 23:56:35.5290  2021-01-01 00:01:03.8460   
19506841           636  2020-12-31 23:56:52.5540  2021-01-01 00:07:28.8400   
19506842          1124  2020-12-31 23:57:15.6080  2021-01-01 00:15:59.8620   
19506843           369  2020-12-31 23:57:17.1180  2021-01-01 00:03:27.1140   
19506844          1060  2020-12-31 23:57:23.8780  2021-01-01 00:15:04.4860   
19506845           184  2020-12-31 23:57:33.1060  2021-01-01 00:00:37.6170   
19506846          1970  2020-12-31 23:57:47.3580  2021-01-01 00:30:38.0560   
19506847           519  2020-12-31 23:57:50.2000  2021-01-01 00:06:29.6850   
19506848           733  2020-12-31 23:58:10.1570  2021-01-01 00:10:23.5950   
19506849          1178  2020-12-31 23:58:14.1000  2021-01-01 00:17:52.3380   
19506850          1344  2020-12-31 23:58:17.4800  2021-01-01 00:20:41.6070   
19506851           589  2020-12-31 23:58:21.2620  2021-01-01 00:08:10.9220   
19506852          2045  2020-12-31 23:58:21.7040  2021-01-01 00:32:27.1570   
19506853           288  2020-12-31 23:58:37.8400  2021-01-01 00:03:26.3250   
19506854           152  2020-12-31 23:58:52.8090  2021-01-01 00:01:25.6720   
19506855           654  2020-12-31 23:59:55.9550  2021-01-01 00:10:50.6050   
19506856          2589  2020-12-31 23:59:57.4070  2021-01-01 00:43:06.4490   

          start_station_id                   start_station_name  \
19506837               464                      E 56 St & 3 Ave   
19506838               312                Allen St & Stanton St   
19506839              3628                 Lenox Ave & W 117 St   
19506840              4099               E 169 St & Webster Ave   
19506841              3143                      5 Ave & E 78 St   
19506842               490                      8 Ave & W 33 St   
19506843              3288                      E 88 St & 1 Ave   
19506844              3977          Prospect Ave & Longwood Ave   
19506845               372            Franklin Ave & Myrtle Ave   
19506846              3169               Riverside Dr & W 82 St   
19506847               265             Stanton St & Chrystie St   
19506848              3143                      5 Ave & E 78 St   
19506849              3335                     Union St & 4 Ave   
19506850               456                E 53 St & Madison Ave   
19506851              3999  Adam Clayton Powell Blvd & W 138 St   
19506852               526                      E 33 St & 5 Ave   
19506853              3307               West End Ave & W 94 St   
19506854              3510  Adam Clayton Powell Blvd & W 123 St   
19506855              4073          W 51 St & Rockefeller Plaza   
19506856              3581           Underhill Ave & Lincoln Pl   

          start_station_latitude  start_station_longitude  end_station_id  \
19506837               40.759345               -73.967597             456   
19506838               40.722055               -73.989111             265   
19506839               40.802557               -73.949078            3137   
19506840               40.834015               -73.908863            4177   
19506841               40.776321               -73.964274            3142   
19506842               40.751551               -73.993934             490   
19506843               40.778301               -73.948813            3282   
19506844               40.818990               -73.902020            3977   
19506845               40.694546               -73.958014             364   
19506846               40.787209               -73.981281             529   
19506847               40.722293               -73.991475             307   
19506848               40.776321       

In [21]:
#Creating a subset of the unique End Stations
NYB2020_StationsUniqueEnds = NYB2020_Stations.drop_duplicates(subset = ["end_station_id"])

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19505857 entries, 0 to 19506856
Data columns (total 29 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   tripduration             int64  
 1   starttime                object 
 2   stoptime                 object 
 3   start_station_id         int16  
 4   start_station_name       object 
 5   start_station_latitude   float64
 6   start_station_longitude  float64
 7   end_station_id           int16  
 8   end_station_name         object 
 9   end_station_latitude     float64
 10  end_station_longitude    float64
 11  bikeid                   int32  
 12  usertype                 object 
 13  birth_year               int16  
 14  gender                   int8   
 15  start_hour               int8   
 16  start_date2              object 
 17  temp                     int8   
 18  rain_snow                float16
 19  day_of_week              object 
 20  day_of_week_number       int8   
 21  month 

In [22]:
NYB2020_StationsUniqueEnds.to_csv(os.path.join(path, '02 Data','Prepared Data', 'NYB2020_UniqueEnds.csv'), index = False)